### PR TITLE
libsubprocess: support FLUX_SUBPROCESS_FLAGS_LOCAL_UNBUF

### PR DESCRIPTION
**Pull request by chu11**
Per discussion in #5919, some notes

- built on top of my cleanups in #5974
- should the use of this flag in  job-exec be a different PR?
- in job-exec subprocess flags are passed in via the API to bulk-exec, so it felt wrong to just assume that flag is always set.  So that led to adding a new `flux_subprocess_get_flags()` function and calling an appropriate "read" function depending on flags.  This could be done differently of course, perhaps the flags aren't passed in from the API and are internalized within `bulk-exec`?
- haven't tested to ensure that the performance profile changes as expected, dunno if we have a simple reproducer laying around? Edit: I did now, see notes in #5919

### CODE REVIEW

File: src/modules/job-exec/bulk-exec.c [hunk 1]
```diff
@@ -222,11 +222,20 @@ static void exec_output_cb (flux_subprocess_t *p, const char *stream)
...
+    int flags = flux_subprocess_get_flags (p);
     int len;
 
-    if ((len = flux_subprocess_getline (p, stream, &s)) < 0) {
-        flux_log_error (exec->h, "flux_subprocess_getline");
-        return;
+    if (flags & FLUX_SUBPROCESS_FLAGS_LOCAL_UNBUF) {
+        if ((len = flux_subprocess_read_local_unbuf (p, stream, &s)) < 0) {
+            flux_log_error (exec->h, "flux_subprocess_read_local_unbuf");
+            return;
```
Comment (garlick) [inline src/modules/job-exec/bulk-exec.c:231]: I think it would be fine to assume the UNBUF flag, e.g. just |= to the flags provided  to` bulk_exec_push_cmd()` and drop the new public function.

**Review by garlick — COMMENTED** [review]
Just a couple of quick comments on a first pass

File: src/modules/job-exec/bulk-exec.c (hunk 1 above)
Comment (chu11) [reply src/modules/job-exec/bulk-exec.c:231]: sounds good

File: src/common/libsubprocess/subprocess.h [hunk 2]
```diff
@@ -265,6 +273,18 @@ int flux_subprocess_getline (flux_subprocess_t *p,
...
+ *  exclusively with FLUX_SUBPROCESS_FLAGS_LOCAL_UNBUF.
+ *
+ *  Unlike flux_subprocess_read() the returned data is not NUL
+ *  terminated.  This can be useful for performance improvements,
+ *  especially when data is written out to storage instead of a user's
+ *  terminal.
+ */
+int flux_subprocess_read_local_unbuf (flux_subprocess_t *p,
+                                      const char *stream,
+                                      const char **bufp);
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libsubprocess/subprocess.h:286]: Sure on another vote.

Documentation that `flux_subprocess_read()` does not NULL terminate seems sufficient to me but maybe I like to live dangerously :-)

Alternatively, mabye we could just change `iodecode()` to add the termination?

[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libsubprocess/subprocess.h:286]: I agree it is fine to change behavior of the `flux_subprocess_read()` function based on the flag.

I haven't re-familiarized myself with the code enough to comment on @garlick's proposal to change `iodecode()` as an alternative though.

[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libsubprocess/subprocess.h:286]: Well if we've got two votes to not worry about changing the behavior then I would say lets not get into changing `iodecode()`, if @chu11 agrees.

**Comment by chu11** [comment]
re-pushed given the comments above, did an additional profiling run and it showed some smalll incremental improvements.

File: src/common/libsubprocess/subprocess.c [hunk 3]
```diff
@@ -794,6 +795,24 @@ static int subprocess_read (flux_subprocess_t *p,
...
+        if (!local_unbuf_allowed) {
+            errno = EPERM;
+            return -1;
+        }
+        /* If user called this function twice, return nothing.  They
+         * will hopefully call flux_subprocess_read_stream_closed() to
+         * verify EOF was not yet reached.
+         */
+        if (c->unbuf_consumed)
+            return 0;
```
Comment (garlick) [inline src/common/libsubprocess/subprocess.c:808]: The flag and the check seem unnecessary.  We can probably just document that reading twice returns the same data and it won't be very surprising given the lack of a buffer.

See doc comment below.

File: src/common/libsubprocess/subprocess.h [hunk 4]
```diff
@@ -61,6 +61,13 @@ enum {
     FLUX_SUBPROCESS_FLAGS_SETPGRP = 2,
     /* use fork(2)/exec(2) even if posix_spawn(3) available */
     FLUX_SUBPROCESS_FLAGS_FORK_EXEC = 4,
+    /* flux_rexec(): In order to improve performance, do not locally
+     * copy and buffer output from the remote subprocess.  Immediately
+     * call output callbacks.  Users can directly retrieve data via
+     * flux_subprocess_read().  Data will not NUL terminated.  Use of
+     * other read functions will result in a EPERM error.
+     */
+    FLUX_SUBPROCESS_FLAGS_LOCAL_UNBUF = 8,
```
Comment (garlick) [inline src/common/libsubprocess/subprocess.h:70]: >   Users can directly retrieve data via `flux_subprocess_read()`.

We probably should specify that `flux_subprocess_read()` must be called exactly once when the output callback is invoked.  If not called, data is lost.  If called more than once, the same data is returned.

> Data will not NUL terminated.

Missing a "be".

> Use of other read functions will result in a EPERM error.

EPERM is kind of strange here.  Would EINVAL be better?

File: src/common/libsubprocess/subprocess.c [hunk 5]
```diff
@@ -829,22 +848,22 @@ int flux_subprocess_read (flux_subprocess_t *p,
                           const char *stream,
                           const char **bufp)
 {
-    return subprocess_read (p, stream, bufp, false, false, false, NULL);
+    return subprocess_read (p, stream, bufp, false, false, false, true, NULL);
```
Comment (garlick) [inline src/common/libsubprocess/subprocess.c:851]: This method of reducing duplicated code with a common function with booleans that alter its behavior is hard to follow.  Instead of adding one more boolean to this function, how about adding a check to `flux_subprocess_read_line()` et al like this:
```c
if (p && (p->flags & FLUX_SUBPROCESS_FLAGS_LOCAL_UNBUF)) {
    errno = EINVAL;
    return -1;
}
```

File: src/common/libsubprocess/remote.c [hunk 6]
```diff
@@ -543,11 +545,63 @@ int subprocess_remote_setup (flux_subprocess_t *p, const char *service_name)
...
+     * seen, we can send it in the local unbuf immediately */
+    if (eof) {
+        /* N.B. we technically should clear/memset c->unbuf_data
+         * before returning EOF, as a user could read from it during
+         * the eof callback.  But for performance reasons, we won't do
+         * that and assume the caller is smart enough not to do that.
+         */
+        c->read_eof_received = true;
+        c->unbuf_len = 0;
+        c->unbuf_eof = true;
```
Comment (garlick) [inline src/common/libsubprocess/remote.c:590]: Why do we have both `c->eof_received` and `c->unbuf_eof`?

File: src/common/libsubprocess/test/remote.c [hunk 7]
```diff
@@ -307,6 +307,106 @@ void simple_test (flux_t *h)
...
+
+    if (!streq (stream, "stdout"))
+        BAIL_OUT ("unexpected stream: %s", stream);
+
+    if (ctx->scorecard.stdout_lines == 0) {
+
+        len = flux_subprocess_read_line (p, stream, &line);
+        ok (len < 0
+            && errno == EPERM,
+            "flux_subprocess_read_line fails w/ EPERM w/ LOCAL_UNBUF");
```
Comment (garlick) [inline src/common/libsubprocess/test/remote.c:325]: Tests like this should zero `errno` before testing that the function sets it.

File: src/modules/job-exec/bulk-exec.c [hunk 8]
```diff
@@ -264,7 +264,8 @@ static struct exec_cmd *exec_cmd_create (const struct idset *ranks,
         fprintf (stderr, "exec_cmd_create: flux_cmd_copy failed");
         goto err;
     }
-    c->flags = flags;
+    /* bulk-exec defaults to use unbuffered for performance */
+    c->flags = flags | FLUX_SUBPROCESS_FLAGS_LOCAL_UNBUF;
```
Comment (garlick) [inline src/modules/job-exec/bulk-exec.c:268]: I'd just say "bulk-exec always uses unbuffered reads for performance" since calling it a default sort of implies that there could be a way to change it.

**Review by garlick — COMMENTED** [review]
I made a first pass and had some comments/questions

File: src/common/libsubprocess/remote.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (chu11) [reply src/common/libsubprocess/remote.c:590]: ahh that comment was leftover from a prior iteration of this PR.  I should remove it.

[the remark this answers is not in the document]
Comment (chu11) [reply src/common/libsubprocess/remote.c:590]: > If setting c->unbuf_len to zero why not set c->unbuf_data to NULL?

I guess it wasn't completely necessary, but you are correct, it wouldn't hurt.

> If the remote returns data + EOF, doesn't setting c->unbuf_len to zero cause the data to be lost?

Within this function, all the data is returned in the first block in the following code.  Then the EOF in the following block.  So the second setting of unbuf_len = 0 is intentional, to clear out that data and send EOF to the caller.

```
    if (data && len) {                                                                                                                      
        c->unbuf_data = data;                                                                                                               
        c->unbuf_len = len;                                                                                                                 
        c->unbuf_consumed = false;                                                                                                          
        if (eof)                                                                                                                            
            c->read_eof_received = true;                                                                                                    
        c->unbuf_eof = false;                                                                                                               
                                                                                                                                            
        c->output_cb (c->p, c->name);                                                                                                       
                                                                                                                                            
        c->unbuf_consumed = false;                                                                                                          
    }                                                                                                                                       
    /* N.B. any data not consumed by the user is lost, so if eof is                                                                         
     * seen, we can send it in the local unbuf immediately */                                                                               
    if (eof) {                                                                                                                              
        c->read_eof_received = true;                                                                                                        
        c->unbuf_data = NULL;                                                                                                               
        c->unbuf_len = 0;                                                                                                                   
        c->unbuf_eof = true;                                                                                                                
                                                                                                                                            
        c->output_cb (c->p, c->name);                                                                                                       
                                                                                                                                            
        c->eof_sent_to_caller = true;                                                                                                       
        c->p->channels_eof_sent++;                                                                                                          
    }                                
```

File: src/common/libsubprocess/subprocess.c (hunk 3 above)
Comment (chu11) [reply src/common/libsubprocess/subprocess.c:808]: > The flag and the check seem unnecessary. We can probably just document that reading twice returns the same data and it won't be very surprising given the lack of a buffer.

I was trying to be consistent to how the buffered case works.  But yeah, we could tweak that.

File: src/common/libsubprocess/subprocess.h (hunk 4 above)
Comment (chu11) [reply src/common/libsubprocess/subprocess.h:70]: > EPERM is kind of strange here. Would EINVAL be better?

I currently use EPERM for another case where it is "illegal" to call a certain function.  It seems an ok choice.  To me EINVAL just makes you think bad input.  Hmmm, a better option isn't coming to me.  Will sleep on it.

File: src/common/libsubprocess/remote.c (hunk 6 above)
Comment (chu11) [reply src/common/libsubprocess/remote.c:590]: I think `unbuf_eof` is left over from a prior attempt.  Should remove it.

[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libsubprocess/remote.c:590]: Whoops!  My bad, I did not catch the `output_cb()` call in there.

File: src/common/libsubprocess/subprocess.h (hunk 4 above)
Comment (garlick) [reply src/common/libsubprocess/subprocess.h:70]: No big deal I suppose!

**Comment by chu11** [comment]
re-pushed w/ tweaks per comments above.  Only notable thing is kept EPERM vs EINVAL for the "can't use these functions" error case.

**Comment by garlick** [comment]
Sounds good.  I'll get this on my test system that is configured for sdexec and see how it goes.

One concern is the lack of output buffering in `sdexec`, which in combination with this flag means no buffering at all.  I'll try it and we should also audit `job-exec` with that in mind to see if we can spot any problem areas.   We may need to add output buffering to `sdexec` before switching `job-exec` over if there are problems.

**Comment by garlick** [comment]
The shell barrier protocol is fine since I believe it's the only user of stdout, and the protocol consists of the shell sending a line of output and then job exec sending one line back.  There is not much opportunity for buffering (or lack thereof) to screw that up.

@grondo suggested I print something from a userrc to check how `job-exec` handles shell output on stderr.

```lua
-- foo.lua
io.stderr:write("foo bar\n")
io.stderr:write("foo bar\n")
io.stderr:write("foo bar\n")
io.stderr:write("foo bar\n")
```
With master
```
$ flux run -o userrc=foo.lua true
0.040s: flux-shell[0]: stderr: foo bar
0.040s: flux-shell[0]: stderr: foo bar
0.040s: flux-shell[0]: stderr: foo bar
0.040s: flux-shell[0]: stderr: foo bar
$ flux job eventlog -H -p exec $(flux job last) | grep log
[  +0.017012] log component="flux-shell" stream="stderr" rank="0" data="foo bar\n"
[  +0.017024] log component="flux-shell" stream="stderr" rank="0" data="foo bar\n"
[  +0.017031] log component="flux-shell" stream="stderr" rank="0" data="foo bar\n"
[  +0.017038] log component="flux-shell" stream="stderr" rank="0" data="foo bar\n"
```

With this PR there is evidence that lines are not being buffered.
Surprisingly, this happens with and without sdexec.  I thought we had remote line buffering with the normal subprocess server.
```
$ flux run -o userrc=foo true
0.064s: flux-shell[0]: stderr: foo bar
foo bar
foo bar
foo bar
$ flux job eventlog -H -p exec $(flux job last) | grep log
[  +0.033395] log component="flux-shell" stream="stderr" rank="0" data="foo bar\nfoo bar\nfoo bar\nfoo bar\n"

```

**Comment by garlick** [comment]
Oh!   Even though the remote subprocess is line buffered (the default), we are using `flux_subprocess_read()` in the server to consume data from the buffer and put it into messages.  We need to use the line oriented read functions so that we are sending only one line per message or else the benefits of line buffering are lost.  I guess the server output callback needs to check the stream's LINE_BUFFER option setting and then use the appropriate read function.

**Comment by garlick** [comment]
This seems to fix that problem.
```diff
diff --git a/src/common/libsubprocess/server.c b/src/common/libsubprocess/server.c
index fca01444e..139923f23 100644
--- a/src/common/libsubprocess/server.c
+++ b/src/common/libsubprocess/server.c
@@ -246,7 +246,10 @@ static void proc_output_cb (flux_subprocess_t *p, const char *stream)
     const char *buf;
     int len;
 
-    if ((len = flux_subprocess_read (p, stream, &buf)) < 0) {
+    len = flux_subprocess_getline (p, stream, &buf);
+    if (len < 0 && errno == EPERM) // not line buffered
+        len = flux_subprocess_read (p, stream, &buf);
+    if (len < 0) {
         llog_error (s,
                     "error reading from subprocess stream %s: %s",
                     stream,
```

**Comment by chu11** [comment]
> This seems to fix that problem.

Hmmmm.  There's a part of me wondering if I originally messed up and had intended to NOT line buffer on the server side, i.e. all line buffer handling occurs on the client side.

But now that we have LOCAL_UNBUF, perhaps that no longer matters and the approach you have is what needs to be done.

**Comment by garlick** [comment]
I guess it's a trade off.  The client has turned out to be pretty heavy with the buffering on that end.  OTOH if we buffer on the server end then we can end up with more message framing overhead when there are bursts of messages.  But since it's a common idiom to have many clients in one place spread out across many servers, moving the buffering to the server side probably makes sense.

Should we add the server side change in this PR and also maybe a test to show that we get one callback per line in UNBUF mode with the libsubprocess server?  Then we could follow up with a PR for sdexec to add some buffering to it.  I can look at that.

**Comment by chu11** [comment]
re-pushed, adding commit to update server side to send each line buffered chunk and added a multiline test.

**Review by garlick — APPROVED** [review]
Seems like we're good!

**Comment by chu11** [comment]
hmmm alpine builder failed

```
Error: not ok 13 - job-exec: can specify default-shell on cmdline
Error: not ok 16 - job-exec: update default shell via config reload
Error: not ok 17 - job-exec: cmdline default shell takes priority
Error: not ok 18 - job-exec: can specify imp path on cmdline
Error: not ok 21 - job-exec: update imp path via config reload
Error: not ok 22 - job-exec: cmdline imp path takes priority
Error: not ok 23 - job-exec: can specify exec service on cmdline
Error: not ok 25 - job-exec: update exec service via config reload
Error: not ok 26 - job-exec: cmdline exec service takes priority
Error: not ok 28 - job-exec: update exec service override via config reload
Error: not ok 32 - job-exec: update sdexec properties via config reload
Error: ERROR: t2403-job-exec-conf.t - exited with status 1
```

these are all job-exec reload module tests.  Wondering if the recent additions kill-timeout, etc. have issues when the module is reloaded.  on alpine, the globals may not be reset.

**Comment by chu11** [comment]
without any better ideas, re-pushed with an extra commit that re-initializes the new signaling globals.  I'm guessing my changes to job-exec tickled things enough to make things fail.

**Comment by grondo** [comment]
Ah, yeah, that makes sense. Probably we were only lucky to get away with it previously.

**Comment by chu11** [comment]
oh whoops, i should have removed `merge-when-passing` until that new commit was reviewed .... hopefully it's ok :-)  here it is

https://github.com/flux-framework/flux-core/pull/5975/commits/a35f615a06693c23a838adf804e93f993e32faa1

**Comment by grondo** [comment]
Well that change overwrites the previous setting on any configuration reload with the defaults. However, probably not a big deal since these are mostly used only for testing at this point...

**Comment by chu11** [comment]
> Well that change overwrites the previous setting on any configuration reload with the defaults. However, probably not a big deal since these are mostly used only for testing at this point...

Gah ... that's right, should have passed the saved argc/argv into it like we do with the exec_config ... I'll do a fix.

```
    if (job_exec_set_config_globals (h, conf, 0, NULL, &err) < 0)
        goto error;

    while ((impl = implementations[i]) && impl->name) {
        if (impl->config) {
            if ((*impl->config) (h, conf, ctx->argc, ctx->argv, &err) < 0)
                goto error;
        }
        i++;
    }
```
